Update EIP-2935: bring up to date with sys contract impl (#9144)
ethereum/EIPs: EIPS/eip-2935.md

| Parameter | Value |
| - | - |
| `BLOCKHASH_SERVE_WINDOW`  | `256` |

> | `HISTORY_SERVE_WINDOW`    | `8192` |

Reviewer: ```suggestion
| `HISTORY_SERVE_WINDOW`    | `8191` |
```

| Parameter | Value |
| - | - |
| `FORK_TIMESTAMP`          | TBD |
| `BLOCKHASH_SERVE_WINDOW`  | `256` |

> | `HISTORY_SERVE_WINDOW`    | `8192` |

Reviewer 2: ```suggestion
| `HISTORY_SERVE_WINDOW`    | `8191` |
```
Update according to PR description, but I have not verified this value.

It is used from the EVM for looking up block hashes.

* Callers provide the block number they are querying in a big-endian encoding.
* If calldata is not 32 bytes, revert.

> * For any request outside the range of [block.number-`HISTORY_SERVE_WINDOW`, block.number-1], revert.

Reviewer: this is not the current BLOCKHASH (with window 256) behavior so we are actually diverting from the behavior ?

Author: mentioned this elsewhere, but I think this divergence is okay because i) it matches other system contracts and ii) it is pretty straightforward to create an adapter around this system contract to mimic the BLOCKHASH behavior when it is used to back the actual op.

Exact evm assembly that can be used for the history contract:

```
caller
push20 0xfffffffffffffffffffffffffffffffffffffffe
eq
push1 0x46
jumpi
push1 0x20
calldatasize
sub
push1 0x42
jumpi

> push0

Author: I removed all these comments because the assembled code comes from the command:

`disease --code 0x$(geas src/consolidations/main.eas) | sed -E '/^[[:space:]]*$/d; s/^[[:space:]]*[0-9a-fA-F]+:[[:space:]]+//; s/[[:space:]]*# selector\(.*\)$//'`

all the code info is in the sysasm: https://github.com/lightclient/sys-asm/blob/main/src/execution_hash/main.eas

I'd probably hold off on code comments in the eip until this is going to last call.

jumpi

// check if input > 8 byte value and revert if this isn't the case
// the check is performed by comparing the biggest 8 byte number with
// the call data, which is a right-padded 32 byte number.
push8 0xffffffffffffffff
push0
calldataload
gt
push1 0x53
jumpi

// check if input > blocknumber-1 then return 0
push1 0x1
push1 0x01
number
sub
push0
calldataload
dup2
gt
push1 0x4b
push1 0x42
jumpi

// check if blocknumber > input + 8192 then return 0, no overflow expected for input of < max 8 byte value
push0
calldataload
push2 0x2000
add
push2 0x1fff
dup2
number
sub
gt
push1 0x4b
push1 0x42
jumpi

// mod 8192 and sload
push2 0x1fff
push0
calldataload
and
swap1
mod
sload

// load into mem and return 32 bytes
push0
mstore
push1 0x20
push0
return

// 0x4b: return 0
jumpdest
push0
push0
mstore
push1 0x20
push0
return

// 0x53: revert
jumpdest
push0
push0
revert

> // 0x57: set op - sstore the input to number-1 mod 8192

Author: I removed all these comments because the assembled code comes from the command:

`disease --code 0x$(geas src/consolidations/main.eas) | sed -E '/^[[:space:]]*$/d; s/^[[:space:]]*[0-9a-fA-F]+:[[:space:]]+//; s/[[:space:]]*# selector\(.*\)$//'`

all the code info is in the sysasm: https://github.com/lightclient/sys-asm/blob/main/src/execution_hash/main.eas

I'd probably hold off on code comments in the eip until this is going to last call.

The first option is recommended until the Verkle fork, to stay consistent with [EIP-4788](./eip-4788.md) and to issues for misconfigured networks where this EIP is activated but history contract hasn't been deployed. The recommendation may be reconsidered at the Verkle fork if filtering the system contract code chunks is deemed too complex.

### Size of ring buffers

> The ring buffer data structure is sized to hold 8191 hashes. In other system contracts a prime ring buffer size is chosen in because using a prime as the modulus ensures that no value is overwritten until the entire ring buffer has been saturated and thereafter, each value will be updated once per iteration, regardless of if some slot are missing or the slot time changes. However, in this EIP the block number is the value in the modulo operation and it only ever increases by 1 each iteration. Which means we can be confident that the ring buffer will always remain saturated.

Reviewer: ```suggestion
The ring buffer data structure is sized to hold 8191 hashes. In other system contracts a prime ring buffer size is chosen in because using a prime modulus ensures that the entire ring buffer gets saturated. However, in this EIP the block number only ever increases by 1 each iteration fully saturating the buffer before wrapping around.
```

> For consistency with other system contracts, we have decided to retain the buffer size of 8191. Given the current mainnet values, 8191 roots provides about a day of coverage. This also gives users plenty of time to make a transaction with a verification against a specific hash and get the transaction included on-chain.

Reviewer: ```suggestion
However for consistency with other system contracts, we have decided to retain the buffer size of 8191. Given the current mainnet values, 8191 roots provides about a day of coverage. This also gives users plenty of time to make a transaction with a verification against a specific hash and get the transaction included on-chain.
```
